src/ray/object_manager/object_manager.cc

```diff
@@ -285,6 +202,60 @@ void ObjectManager::PullSendRequest(const ObjectID &object_id,
       });
 }
 
+void ObjectManager::RestartPullTimer(const ObjectID &object_id) {
+  auto timer = boost::asio::deadline_timer(*main_service_);
+  boost::posix_time::milliseconds retry_timeout(
+      RayConfig::instance().object_manager_pull_timeout_ms());
+  timer.expires_from_now(retry_timeout);
+  timer.async_wait([this, object_id](const boost::system::error_code &error) {
+    if (!error) {
+      std::vector<ClientID> clients_to_request;
+      bool abort_creation;
+      bool restart_timer;
+      pull_manager_.TimerExpired(object_id, &clients_to_request, &abort_creation,
+                                 &restart_timer);
+      if (abort_creation) {
+        AbortObjectCreation(object_id);
+      } else {
+        for (const ClientID &client_id : clients_to_request) {
+          SendPullRequest(object_id, client_id);
+        }
+      }
+      if (restart_timer) {
+        RestartPullTimer(object_id);
+      } else {
+        // If we do not need to restart the timer, we remove the timer from
+        // pull_timers.
+        pull_timers_.erase(object_id);
+      }
+    } else {
+      // Check that the error was due to the timer being canceled.
+      RAY_CHECK(error == boost::asio::error::operation_aborted);
+    }
+  });
+
+  // This will cancel and replace the old timer if there was an old one.
+  auto it = pull_timers_.find(object_id);
+  if (it == pull_timers_.end()) {
+    auto insertion_it = pull_timers_.insert(std::make_pair(object_id, std::move(timer)));
+    RAY_CHECK(insertion_it.second);
+  } else {
+    it->second = std::move(timer);
+  }
+}
+
+void ObjectManager::AbortObjectCreation(const ObjectID &object_id) {
+  // TODO(rkn): Implement this!
+
+  RAY_LOG(WARNING) << "We do not handle the abort_creation case yet.";
```

Comment:
this would be implemented in a later PR?

Reply:
yes i believe thats the plan